extractTimeRange fails to extract time lower bound when column is on the right side

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

 1. In `pkg/planner/core/operator/logicalop/logicalop_test/logical_mem_table_predicate_extractor_test.go`, add this case to `TestTiDBHotRegionsHistoryTableExtractor`:

  ```go
  {
  	sql:            "select * from information_schema.tidb_hot_regions_history where '2019-10-10 10:10:10' < update_time",
  	startTime:      timestamp(t, "2019-10-10 10:10:10") + 1,
  	isLearners:     []bool{false, true},
  	isLeaders:      []bool{false, true},
  	hotRegionTypes: set.NewStringSet(plannercore.HotRegionTypeRead, plannercore.HotRegionTypeWrite),
  },
```

  2. Run:
```
  go test ./pkg/planner/core/operator/logicalop/logicalop_test -run TestTiDBHotRegionsHistoryTableExtractor -count=1
```

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

The predicate extractor should treat `'2019-10-10 10:10:10' < update_time` as equivalent to `update_time > '2019-10-10 10:10:10'` and extract the lower bound correctly.

For this test case, StartTime should be 1570673410001.

### 3. What did you see instead (Required)

The time range was not extracted when the column appeared on the right-hand side of the comparison.

StartTime stayed 0, and the test failed with:

```
--- FAIL: TestTiDBHotRegionsHistoryTableExtractor
    logical_mem_table_predicate_extractor_test.go:1446:
        Error:       Not equal:
                     expected: 1570673410001
                     actual  : 0
        Messages:    SQL: select * from information_schema.tidb_hot_regions_history where '2019-10-10 10:10:10' < update_time
```

This suggests extractTimeRange in pkg/planner/core/memtable_predicate_extractor.go does not correctly handle comparison predicates where the target column is on the right side.

### 4. What is your TiDB version? (Required)

Current source checkout:
- Git Commit Hash: 1c3f9eb211a37aabe976834547b350483918863f
- Git Branch: master

<!-- Paste the output of SELECT tidb_version() -->

--- a/pkg/planner/core/memtable_predicate_extractor.go
+++ b/pkg/planner/core/memtable_predicate_extractor.go
@@ -183,7 +183,7 @@ func (helper *extractHelper) extractColBinaryOpConsExpr(
 	ctx base.PlanContext,
 	extractCols map[int64]*types.FieldName,
 	expr *expression.ScalarFunction,
-) (string, []types.Datum) {
+) (string, []types.Datum, bool) {
 	args := expr.GetArgs()
 	var col *expression.Column
 	var colIdx int
@@ -203,12 +203,12 @@ func (helper *extractHelper) extractColBinaryOpConsExpr(
 		col, colIdx = innerCol, innerColIdx
 	}
 	if col == nil {
-		return "", nil
+		return "", nil, false
 	}
 
 	name, found := extractCols[col.UniqueID]
 	if !found {
-		return "", nil
+		return "", nil, false
 	}
 
 	// The `lhs/rhs` of EQ expression must be a constant
@@ -216,7 +216,7 @@ func (helper *extractHelper) extractColBinaryOpConsExpr(
 	// SELECT * FROM t1 WHERE 'lhs'=c
 	constant, ok := args[1-colIdx].(*expression.Constant)
 	if !ok || constant.DeferredExpr != nil {
-		return "", nil
+		return "", nil, false
 	}
 	v := constant.Value
 	if constant.ParamMarker != nil {
@@ -225,10 +225,10 @@ func (helper *extractHelper) extractColBinaryOpConsExpr(
 		intest.AssertNoError(err, "fail to get param")
 		if err != nil {
 			logutil.BgLogger().Warn("fail to get param", zap.Error(err))
-			return "", nil
+			return "", nil, false
 		}
 	}
-	return name.ColName.L, []types.Datum{v}
+	return name.ColName.L, []types.Datum{v}, colIdx == 0
 }
 
 // extract the OR expression, e.g:
@@ -247,7 +247,8 @@ func (helper *extractHelper) extractColOrExpr(ctx base.PlanContext, extractCols 
 	var extract = func(extractCols map[int64]*types.FieldName, fn *expression.ScalarFunction) (string, []types.Datum) {
 		switch helper.getStringFunctionName(fn) {
 		case ast.EQ:
-			return helper.extractColBinaryOpConsExpr(ctx, extractCols, fn)
+			colName, datums, _ := helper.extractColBinaryOpConsExpr(ctx, extractCols, fn)
+			return colName, datums
 		case ast.LogicOr:
 			return helper.extractColOrExpr(ctx, extractCols, fn)
 		case ast.In:
@@ -319,7 +320,7 @@ func (helper *extractHelper) extractCol(
 		switch helper.getStringFunctionName(fn) {
 		case ast.EQ:
 			helper.enableScalarPushDown = true
-			colName, datums = helper.extractColBinaryOpConsExpr(ctx, extractCols, fn)
+			colName, datums, _ = helper.extractColBinaryOpConsExpr(ctx, extractCols, fn)
 			if colName == extractColName {
 				helper.setColumnPushedDownFn(colName, extractCols, fn)
 			}
@@ -449,7 +450,7 @@ func (helper extractHelper) extractLikePattern(
 	var datums []types.Datum
 	switch fn.FuncName.L {
 	case ast.EQ, ast.Like, ast.Ilike, ast.Regexp, ast.RegexpLike:
-		colName, datums = helper.extractColBinaryOpConsExpr(ctx, extractCols, fn)
+		colName, datums, _ = helper.extractColBinaryOpConsExpr(ctx, extractCols, fn)
 	}
 	if colName != extractColName {
 		return false, ""
@@ -551,13 +552,27 @@ func (helper extractHelper) extractTimeRange(
 
 		var colName string
 		var datums []types.Datum
+		var colOnLeft bool
 		fnName := helper.getTimeFunctionName(fn)
 		switch fnName {
 		case ast.GT, ast.GE, ast.LT, ast.LE, ast.EQ:
-			colName, datums = helper.extractColBinaryOpConsExpr(ctx, extractCols, fn)
+			colName, datums, colOnLeft = helper.extractColBinaryOpConsExpr(ctx, extractCols, fn)
 		}
 
 		if colName == extractColName {
+			if !colOnLeft {
+				switch fnName {
+				case ast.GT:
+					fnName = ast.LT
+				case ast.GE:
+					fnName = ast.LE
+				case ast.LT:
+					fnName = ast.GT
+				case ast.LE:
+					fnName = ast.GE
+				}
+			}
+
 			timeType := types.NewFieldType(mysql.TypeDatetime)
 			timeType.SetDecimal(6)
 			timeDatum, err := datums[0].ConvertTo(ctx.GetSessionVars().StmtCtx.TypeCtx(), timeType)

--- a/pkg/planner/core/operator/logicalop/logicalop_test/logical_mem_table_predicate_extractor_test.go
+++ b/pkg/planner/core/operator/logicalop/logicalop_test/logical_mem_table_predicate_extractor_test.go
@@ -443,6 +443,13 @@ func TestClusterLogTableExtractor(t *testing.T) {
 			endTime:   timestamp(t, "2019-10-11 10:10:10") - 1,
 		},
 		{
+			sql:       "select * from information_schema.cluster_log where '2019-10-10 10:10:10' < time and '2019-10-11 10:10:10' > time",
+			nodeTypes: set.NewStringSet(),
+			instances: set.NewStringSet(),
+			startTime: timestamp(t, "2019-10-10 10:10:10") + 1,
+			endTime:   timestamp(t, "2019-10-11 10:10:10") - 1,
+		},
+		{
 			sql:         "select * from information_schema.cluster_log where time>='2019-10-12 10:10:10' and time<'2019-10-11 10:10:10'",
 			nodeTypes:   set.NewStringSet(),
 			instances:   set.NewStringSet(),
@@ -631,6 +638,12 @@ func TestMetricTableExtractor(t *testing.T) {
 			promQL:    `histogram_quantile(0.9, sum(rate(tidb_server_handle_query_duration_seconds_bucket{}[60s])) by (le,sql_type,instance))`,
 			startTime: parseTime(t, "2019-10-10 10:10:10"),
 			endTime:   parseTime(t, "2019-10-10 10:20:10"),
+		},
+		{
+			sql:       "select * from metrics_schema.tidb_query_duration where '2019-10-10 10:10:10' < time and '2019-10-11 10:10:10' > time",
+			promQL:    `histogram_quantile(0.9, sum(rate(tidb_server_handle_query_duration_seconds_bucket{}[60s])) by (le,sql_type,instance))`,
+			startTime: parseTime(t, "2019-10-10 10:10:10.001"),
+			endTime:   parseTime(t, "2019-10-11 10:10:09.999"),
 		},
 		{
 			sql:         "select * from metrics_schema.tidb_query_duration where time>='2019-10-10 10:10:10' and time<='2019-10-09 10:10:10'",
@@ -1138,6 +1151,14 @@ func TestTiDBHotRegionsHistoryTableExtractor(t *testing.T) {
 		{
 			sql:            "select * from information_schema.tidb_hot_regions_history where update_time>='2019-10-10 10:10:10' and update_time<'2019-10-11 10:10:10'",
 			startTime:      timestamp(t, "2019-10-10 10:10:10"),
+			endTime:        timestamp(t, "2019-10-11 10:10:10") - 1,
+			isLearners:     []bool{false, true},
+			isLeaders:      []bool{false, true},
+			hotRegionTypes: set.NewStringSet(plannercore.HotRegionTypeRead, plannercore.HotRegionTypeWrite),
+		},
+		{
+			sql:            "select * from information_schema.tidb_hot_regions_history where '2019-10-10 10:10:10' < update_time and '2019-10-11 10:10:10' > update_time",
+			startTime:      timestamp(t, "2019-10-10 10:10:10") + 1,
 			endTime:        timestamp(t, "2019-10-11 10:10:10") - 1,
 			isLearners:     []bool{false, true},
 			isLeaders:      []bool{false, true},

